### PR TITLE
Use self-hosted runner for ARM64 integration tests

### PR DESCRIPTION
**Pull request by alpeb**
Use self-hosted runner for ARM64 integration tests

This moves the "ARM64 integration tests" job in `relase.yaml` to a
separate custom action that is run in an ARM self-hosted runner tagged
with `[self-hosted, Linux, ARM64]`, tied to the linkerd github org.

We no longer use a local (linux/x86_64) linkerd CLI that connects to an
existing k3s instance in the host. Instead, we run the CLI ARM64 binary
in the host itself, after creating the cluster with k3d (which gets
always torn down at the end of the tests regardless of their success).

This action can be easily triggered and tested through a separate
workflow like this one:

```yaml
name: eraseme - just for testing the arm64-integration action

on: push

jobs:
  arm64_integration_tests:
    runs-on: [self-hosted, Linux, ARM64]
    timeout-minutes: 30
    steps:
      - uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8
      - uses: ./.github/actions/arm64-integration
        with:
          tag: edge-22.11.1
          docker-registry: ghcr.io/linkerd
          k3d-version: v5.4.4

```

The results for that workflow test are here:
https://github.com/linkerd/linkerd2/actions/runs/3464532975/jobs/5786211951
(run took 2m20s to complete)

Please check the "ARM CI host at Equinix Metal" doc in Notion for the
host setup.

## Other Changes

- The cni test was removed.
- Replaced `"$bindir"/docker` with just `docker` in `bin/image-load` as
  we do elsewhere.
- Properly detect k3d arch in `bin/k3d`

### CODE REVIEW

File: .github/actions/arm64-integration/action.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+name: ARM64 integration tests
```
Comment (olix0r) [inline .github/actions/arm64-integration/action.yml:1]: Is there anything arm specific about this action? Just the path to the CLI?

If this isn't actually reusable, we probably should NOT extract this into an action. For one, we won't get dependabot updates to the actions used here.

File: .github/actions/arm64-integration/action.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    - uses: azure/setup-kubectl@7f7e5ba5ea3e491b92e6e8e5819963f8f3a1f076
+    - shell: bash
+      env:
+        TAG: ${{ inputs.tag }}
+        RUN_ARM_TEST: 1
+        LINKERD_DOCKER_REGISTRY: ${{ inputs.docker-registry }}
+      run: bin/tests --name deep --images preload "$CMD"
+    - shell: bash
+      run: bin/k3d cluster delete --all
+      if: always()
```
Comment (olix0r) [inline .github/actions/arm64-integration/action.yml:34]: Why is this necessary? Does the worker not automatically clean up resources created by the action?

File: .github/actions/arm64-integration/action.yml (hunk 1 above)
Comment (alpeb) [reply .github/actions/arm64-integration/action.yml:1]: Ok, this only facilitated testing by calling from a temporary separate workflow. I'll move it back into `release.yaml` if everything else gets a :heavy_check_mark:

File: .github/actions/arm64-integration/action.yml (hunk 2 above)
Comment (alpeb) [reply .github/actions/arm64-integration/action.yml:34]: Nope, this runs directly into the server and no automatic cleanup is performed.
`bin/tests` by default cleans up the cluster but not when the suite fails.
Tested in 0ce4351

Comment (alpeb) [reply .github/actions/arm64-integration/action.yml:34]: To further clarify, there's no container/vm setup here that would make things transient.

Comment (kleimkuhler) [reply .github/actions/arm64-integration/action.yml:34]: Do we want to consider keeping the cluster around when tests fail? Flakiness could be annoying, but if they are failing consistently it could be nice to be able to SSH into the runner for further debugging.

Comment (alpeb) [reply .github/actions/arm64-integration/action.yml:34]: @kleimkuhler What do you think about having the cleanup happen before the tests get triggered? If the tests succeed they'll get cleaned up by default anyway, if they fail one would be able to see what happened and the cluster would get cleaned before the next run.
Note though that a lot of tests are wrapped around `TestHelper.WithDataPlaneNamespace` which kills the namespace when the tests finishes (successfully or not). I think it'd be a nice improvement (outside this PR) to bypass the killing when the test fails.

Comment (kleimkuhler) [reply .github/actions/arm64-integration/action.yml:34]: So we cleanup old clusters before running a new set of tests? That sounds good to me. If things fail then we'll have time to debug further if we need to.

And yep it's unfortunate we cleanup Namespaces with that function even if the tests fail, but I agree it's better to sort that out separately!

Comment (olix0r) [reply .github/actions/arm64-integration/action.yml:34]: Eh, I'd prefer that we cleanup by default. If you're in the situation where you want to rerun it you can always modify the workflow to disable the cleanup.

Can we randomize the cluster name so that we have no possibility of reusing the same cluster? Or would that require test changes?

Comment (alpeb) [reply .github/actions/arm64-integration/action.yml:34]: > Can we randomize the cluster name so that we have no possibility of reusing the same cluster? Or would that require test changes?

The cluster receives the name of its test, in this case `deep`. We'd have to change the testing library (`bin/_test-helpers.sh`) to override that, perhaps through a flag.

Comment (alpeb) [reply .github/actions/arm64-integration/action.yml]: That sounds good to me @mateiidavid . I've pushed d48b21b using just recipies and the required setup. I've also replaced the CLUSTER_NAME snippet with the branch/tag name as discussed off-github. Note that once the PR gets accepted I'll move this in to `release.yaml` to minimize the chance something's wrong.

**Review by kleimkuhler — APPROVED** [review]
This looks good. Aside from moving back to `release.yml`, we'll also need to make sure `eraseme.yml` is erased.

File: .github/actions/arm64-integration/action.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        echo "PATH=$PATH" >> "$GITHUB_ENV"
+        bin/scurl -v "https://raw.githubusercontent.com/k3d-io/k3d/${{ inputs.k3d-version }}/install.sh" \
+          | USE_SUDO=false K3D_INSTALL_DIR=$PWD/target/bin bash
+    - name: Cluster setup
+      shell: bash
+      run: |
+       CLUSTER_NAME=$(printf "${{ github.ref_name }}" | tr -c [:alnum:] -)
+       echo "CLUSTER_NAME=$CLUSTER_NAME" >> "$GITHUB_ENV"
+       just k3d-name=$CLUSTER_NAME k3d-create
+       just k3d-name=$CLUSTER_NAME k3d-use
```
Comment (mateiidavid) [inline .github/actions/arm64-integration/action.yml:43]: We create the cluster and then in the next step we run the integration tests. At this point in the CI, when the arm tests run, we won't have the images pushed to our registry though, right?

We might have to load the images after `k3d-use`.  Luckily, I think we could merge these two lines, and load the images, by calling into a different recipe altogether.

In the justfile, we set the tag to the output of `bin/tag` -- for a stable release, this should accurately set the stable tag (?)

https://github.com/linkerd/linkerd2/blob/de73ffaa0bc95db83c7b66d7a19b315f48be7a14/justfile#L263

In the `linkerd-load` recipe, we use the tag to load the images:
https://github.com/linkerd/linkerd2/blob/de73ffaa0bc95db83c7b66d7a19b315f48be7a14/justfile#L328-L333

`_linkerd-images` runs first, if we can access the images on the host with `docker ls` (and we should afaik since we do something similar for our normal integration tests, if not we might have to do `docker load` first with tarballs) then all should be well. `_k3d-init` will run second and will create our cluster before the images are loaded. Bonus is, it will also merge the config :)

https://github.com/linkerd/linkerd2/blob/de73ffaa0bc95db83c7b66d7a19b315f48be7a14/justfile#L218-L232

So we could just change this to:
```suggestion
       just k3d-name=$CLUSTER_NAME linkerd-load
```

Or perhaps if we want to be more explicit, create it first then call `linkerd-load`. Do you think this makes sense? Apologies in advance if we are loading the images and I've missed it.

**Review by mateiidavid — COMMENTED** [review]
Most of it looks good except Kevin's comment and some questions around loading the images in the cluster.

File: .github/actions/arm64-integration/action.yml (hunk 3 above)
Comment (alpeb) [reply .github/actions/arm64-integration/action.yml:43]: Thanks for the detailed write-up! The idea is to have this run in the `release.yml` workflow, replacing the current `arm64_integration_tests` job which depends on `docker_build`, so the images are already published at that point. And IMO this test should exercise having the cluster itself pulling the published images. The other non-arm job `integration_tests` does preload the images, but I think that was added for performance, but the ARM host is pretty swift so it shouldn't need that.

Comment (mateiidavid) [reply .github/actions/arm64-integration/action.yml:43]: Ok, makes sense 👍🏻

**Review by mateiidavid — APPROVED** [review]
Good to go 🚢  will we still move the logic in the `release.yml` workflow instead of having it as an action?

**Comment by alpeb** [comment]
> Good to go ship will we still move the logic in the `release.yml` workflow instead of having it as an action?

Yep!
